### ISSUE TITLE
Support multiple formulation definitions in input

### ISSUE THREAD

**Issue by hellkite500** [NOAA-OWP/ngen#187]
Need to support multiple possible formulations per catchment in input configuraiton.

## Current behavior
A catchment is connected to formulation in the following fashion
```json
  "catchments": {
        "test-catchment": {
                "tshirt": { ... },
                "forcing": { ... },
                "other": { ... }
```
## Expected behavior
A list of formulations would be more appropriate, and enforcing some structure on the formulation makes then more generalizable and usable in tool chains (i.e calibration).  Something like:

```json
  "catchments": {
        "test-catchment": {
                "formulations": [
                       { "name": "tshirt", "params": { ... } },
                       { "name": "lumped", "params": { ... } }
                ],
                "forcing": {...},
                "other": {...}
```

Even if we only support a list of size one initially, this provides the base mechanics to easily expand.

**Comment by aaraney** [comment]
Question, why are parameters treated differently from forcings? There is a logical difference between the two, but can we gain anything by considering model input as a layer itself which is inclusive of params, forcings, and things of the like?

**Comment by hellkite500** [comment]
@aaraney parameters have some fundamentally different properties:

1. Parameters are unique to a given model formulation, forcing is not
2. Parameters are _relatively_ *static* at run time, forcings vary in time throughout the runtime
3. Parameters may be subject to estimation/calibration, forcings are not.

These are the three big differences that come to mind off the top of my head.  Parameters and forcings are two different types of inputs that require different types of handling.  We might be able to nudge them together in a more conceptual "input layer" but they would still need significantly different handling/processing both in tool chains pre/post processing data, as well as during model runtimes.

**Comment by aaraney** [comment]
@hellkite500 if we are to think of parameters as we do now with a model like the NWM, I for the most part agree with you points. However, I would argue that fundamentally they are more similar than different, in a concrete imperative sense their differences arise, but that is only due to context which can be handled for example by a given models input engine or something of the like. 

For clarity sake I will response as you have them listed above:
1. If we think about parameters as a space and forcings as a space they are not overly different. While forcing may have extra dimensions (time for example), in many models both parameters and forcings are spatially distributed estimate of past, current, or estimated phenomena. A given model has a subset of all possible parameters in the parameter space (Mannings N, ET rate etc.). Meaning models differ based on their formulation (like you were saying)-- how they represent a process, the inputs they require, spatial-temporal discretization, and numerical solvers implemented. In that vein of thought, parameters are not unique to a given model formulation, the context in which a model uses to ingest and represent a parameter _is_ (usually) unique. The same can be said for forcings, some models require a different forcings and the context for which those forcings are ingests _is_ model dependent.
2. I agree that parameters typically, at run time, have a singular state and retain that state throughout the entire model run. However, you can usually say the same about forcings. They may have been adjusted prior to a runtime, but at run time they are _typically_ static, although they convey units over a given time period.
3. Right, I agree, parameters are commonly used for calibration purposes and likely are estimates. I think it's fair to include forcings as an estimated parameter or set of parameters in a model. This might be a reach, but just trying to keep it general here.

I don't disagree with the points that you made in a concrete imperative sense, however like you mentioned I think at least thinking about a conceptual "input layer" has promising benefits. Is it possible abstract an input layer that models use? Can models provide a context that is used to transform a generalized input (both parameters and forcings) into usable information?

**Comment by dblodgett-usgs** [comment]
Good conversation above... I'll offer a slightly different distinction based on their role in the framework. Hopefully this helps get back to the point of this issue ;).  

- **Forcings** are things that a given formulation must be able to take as input (if supported). 
- **Parameters** are things that a given formulation must be provided to operate that may be unique.
- **Outputs** are things that a given formulation must return when requested (if supported).

There may be some logic in handling dynamic parameters similarly to forcings, but that's an implementation detail.

I think there may be some set of "standard" parameters that could be specified in a non-formulation-specific area of the configuration (area, slope, etc.) but that may be a YAGNI.

**Comment by ZacharyWills** [comment]
As a layman, Forcings and Parameters both seem required to run the model. The difference seems to be in at the interval at which they might change with regards to the objective of the user.

Example 1: 
an Analysis of Record run, you might have (a lot) of fixed **forcings** based on the observed data we have from the historical record, given this you could change the _parameters_ to see their effect on the forecasts in the long term. 

Example 2:
The **parameters** might be fixed, but we might want to ensemble the _forcings_ given that we don't know where the rain will fall.

Given that the temporal nature seems to be what people care about with regards to inputting these things as a user, perhaps there should be some uniformity based on whether or not you want to give **one** value (parameter or forcing) to the whole run or whether you want to give **multiple** values (again forcing or parameter). 

This was mostly me trying to find the line, so if I'm off base certainly let me know.

**Comment by aaraney** [comment]
@hellkite500 @jmframe @dblodgett-usgs 

@ZacharyWills said it much better than I. These points are what I was trying to highlight.
>Example 1:
an Analysis of Record run, you might have (a lot) of fixed forcings based on the observed data we have from the historical record, given this you could change the parameters to see their effect on the forecasts in the long term.
Example 2:
The parameters might be fixed, but we might want to ensemble the forcings given that we don't know where the rain will fall.

**Comment by dblodgett-usgs** [comment]
This temporal variation distinction is true for a given formulation but I think the more fundamental distinction with regard to the framework is around what the framework requires a formulation to do and what a formulation requires the framework provide. That's why I offer the distinction above and suggest that schema and dimensionality of I/O data is kind of secondary (must be flexible).

**Comment by ZacharyWills** [comment]
I understand the distinction here a bit better now, let me see if I can get this right.

There are elements of the exchange between the model formation and the framework that need to be discussed and fleshed out. 

There is also some discussion of how to do that to meet the needs of different kinds of formulations that might rely on the framework.

I think the distinction between an _element_ (defined here as: a required exchange between model and framework) and the _options_ that the element can be configured with for any given formulation. 

So I think once all elements are defined, their options should also be defined with regards to things like how many parameters and inputs you can have and how they're dolled out etc.

**Comment by hellkite500** [comment]
The pure logical support for the OP is implemented in #204
